test/unit/org/apache/cassandra/utils/memory/MemtableCleanerThreadTest.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.utils.memory;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import org.apache.cassandra.metrics.MemtableFlushMetrics;
+import org.apache.cassandra.utils.FBUtilities;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.when;
+
+public class MemtableCleanerThreadTest
+{
+    private static final long TIMEOUT_SECONDS = 5;
+    private static final long TIMEOUT_MILLIS = TimeUnit.SECONDS.toMillis(TIMEOUT_SECONDS);
+
+    @Mock
+    MemtablePool pool;
+
+    @Mock
+    MemtableCleaner cleaner;
+
+    MemtableCleanerThread<MemtablePool> cleanerThread;
+
+    MemtableFlushMetrics metric = new MemtableFlushMetrics();
+
+    @Before
+    public void setup()
+    {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    private void startThread()
+    {
+        cleanerThread = new MemtableCleanerThread<>(pool, cleaner);
+        assertNotNull(cleanerThread);
+        cleanerThread.start();
+
+        for (int i = 0; i < TIMEOUT_MILLIS && !cleanerThread.isAlive(); i++)
+            FBUtilities.sleepQuietly(1);
+
+        assertTrue(cleanerThread.isAlive());
+    }
+
+    private void stopThread()
+    {
+        cleanerThread.shutdownNow();
+
+        for (int i = 0; i < TIMEOUT_MILLIS && cleanerThread.isAlive(); i++)
+            FBUtilities.sleepQuietly(1);
+
+        assertFalse(cleanerThread.isAlive());
+    }
+
+    private void waitForPendingTasks()
+    {
+        // wait for a bit because the cleaner latch completes before the pending tasks are decremented
+        FBUtilities.sleepQuietly(1);
+
+        for (int i = 0; i < TIMEOUT_MILLIS; i++)
+            FBUtilities.sleepQuietly(1);
```

Comment:
Given we don't seem to be checking for the number of pending tasks to short circuit anyway, should we just wait for `TIMEOUT_MILLIS` once and then make sure the pending tasks are done? (We could put the check in the loop if we didn't want to guarantee a wait of that length, but a granularity of 1 millisecond seems too much?)